src/Microsoft.AspNetCore.Blazor/Components/Provider.cs

```diff
@@ -9,7 +9,7 @@ namespace Microsoft.AspNetCore.Blazor.Components
     /// <summary>
     /// A component that provides one or more parameter values to all descendant components.
     /// </summary>
-    public class Provider : IComponent
+    public class Provider<T> : ProviderBase, IComponent
```

Comment:
What's the advantage of turning this into a generic? Is the idea to make it based on declared type rather than runtime type?

Reply:
My only concern is about cases that don't work super well as generics - like delegate conversions

Reply:
> Is the idea to make it based on declared type rather than runtime type?

Yes. The key scenario I have in mind is if you're doing `<Provider Value="someObject">`, but `someObject` is sometimes `null`. If we did it based on runtime type, then the "overriding upstream values" logic would be very strange: it would override if it was non-null, but would not override if it was null. In fact there would be no possible way to override a non-null value with a null one. Given the way we identify the `Provider` just once per tree parameter usage, super bizarre things would happen whereby if the *initial* value was null, you wouldn't be able to observe it become non-null.

Reply:
BTW it was really impressive how making the component generic just worked in every way I could think of (type inference, etc.).